Spacecraft sends notifications when certain events occur

CodeMartian opened the pull request:
#19 

**CHANGELOG** :memo:

* adds a SpacecraftEmitter that will send out messages depending on the following situations:
  -  max_speed: spacecraft attempts to accelerate faster than 5
  - min_speed: spacecraft attempts to decelerate slower than 1
  - wrong_trajectory: spacecraft traverses to the left or right farther than 5
  - moon: the spacecraft reaches the moon
  - lost: the spacesraft passes by the moon

**PRINTS(OPTIONAL)** :framed_picture:
**Spacecraft going too fast and going off course at the same time**
![Spacecraft events](https://user-images.githubusercontent.com/49497114/95022876-eda5d180-0636-11eb-8b59-7e29be3e82d2.png)

**Spacecraft going at the minimum speed, finding the moon, and passing the moon**
![Spacecraft events 2](https://user-images.githubusercontent.com/49497114/95022878-f1395880-0636-11eb-9d24-af8729106eb1.png)


**LABELS**
* Enhancement
* Hacktoberfest

Commit:
add events to spacecraft

Commit:
spacecraft reports events that occur during the journey

The spacecraft now reports any time the following situations happen:
- max_speed: spacecraft attempts to accelerate faster than 5
- min_speed: spacecraft attempts to decelerate slower than 1
- wrong_trajectory: spacecraft traverses to the left or right farther
than 5
- moon: the spacecraft reaches the moon
- lost: the spacesraft passes by the moon

CodeMartian committed:
removes watch mode from build process

CodeMartian committed:
adds new lines after each output

CodeMartian committed:
all of the spacecrafts events notify at once

The spacecraft will now send all of the events that apply to its
situation.

CodeMartian commented:
Okay! @pyymenta I believe this PR is ready for your review. I did some fixin' up and polishing after I created the PR, but I believe it's in a good spot now. I'd love to hear your feedback!

pyymenta commented:
@CodeMartian Thank you for your contribuition, I'll review it!

pyymenta commented on package.json:
```diff
@@ -31,6 +31,7 @@
   "scripts": {
     "start": "node dist/main.js",
     "build": "tsc --build .tsconfig.json",
+    "test:dev": "jest --no-cache --watchAll",
```
Watch command, TDD MODE :on: 

:rocket: :rocket: :rocket:

CodeMartian replied:
🥳

pyymenta commented on src/main.ts:
```diff
@@ -1,8 +1,9 @@
+import SpacecraftEmitter from './spacecraftEmitter';
 
-const spacecraft = new SpaceCraft(Earth, Moon);
+const spacecraft = new SpaceCraft(Earth, Moon, new SpacecraftEmitter(process.stdout));
```
I think you could declare the SpacecraftEmitter in a const and then, pass the instance to the SpaceCraft:

```suggestion
const spaceCraftEmitter = new SpacecraftEmitter(process.stdout);
const spacecraft = new SpaceCraft(Earth, Moon, spaceCraftEmitter);
```

pyymenta commented on src/spacecraftEmitter.ts:
```diff
@@ -0,0 +1,22 @@
+import { EventEmitter } from 'events';
+import { event_messages } from './events_messages.json';
```
Used the @danieljo09 list events :rocket: :rocket: :rocket: :rocket:

pyymenta commented on src/spacecraftEmitter.ts:
```diff
@@ -0,0 +1,22 @@
+export default class SpacecraftEmitter {
+	private eventEmitter: EventEmitter = new EventEmitter();
+	constructor(output: NodeJS.WriteStream) {
+		event_messages.forEach((message) => {
+			this.eventEmitter.on(message.id, () => {
+				output.write(` ${message.message} `);
+			});
+		});
```
I believe it should be a method, to keep the constructor clean, with only a function call. What do you think?

pyymenta commented on src/spacecraftEmitter.ts:
```diff
@@ -0,0 +1,22 @@
+	report(event: string) {
+		const event_message = event_messages.find((message) => {
+			return message.id === event;
+		});
+		if (event_message) {
```
Could you add a space between these commands?

pyymenta commented:
In general your implementation looks good to me, if you fix those points, I believe that will be ready to be merged.

One point of attention is about the identation, I think 4 spaces are not good. I'm thinking to changing it to 2. What do you think? No worry about it now, I'll create another issue for it.

CodeMartian committed:
Update src/main.ts

Co-authored-by: Lucas Pimenta <lucas.pimenta94@outlook.com>

CodeMartian committed:
refactor SpacecraftEmitter to use init function

The logic in the constructor has been extracted into a new private
function called init()

CodeMartian committed:
Merge branch 'master' of https://github.com/CodeMartian/spacecraft-cli

CodeMartian committed:
oops! forgot the space line

CodeMartian committed:
Merge branch 'master' into master

pyymenta reviewed (approved):
Very good @CodeMartian!! You did it! Could I merge in? Or you still working in it?

CodeMartian commented:
go ahead and 🚀 🚀 🚀 🚀 🚀

pyymenta commented:
> go ahead and

Awesome! Thank you!
